core/src/main/java/hudson/model/DirectoryBrowserSupport.java

```diff
@@ -436,33 +437,79 @@ private static void zip(StaplerResponse rsp, VirtualFile root, VirtualFile dir,
-                    //      Unfortunately, only the file separator char of the master is known (File.separatorChar)
-                    //      but not the file separator char of the (maybe remote) "dir".
-                    ZipEntry e = new ZipEntry(relativePath.replace('\\', '/'));
-                    VirtualFile f = dir.child(n);
-                    e.setTime(f.lastModified());
-                    zos.putNextEntry(e);
-                    try (InputStream in = f.open()) {
-                        IOUtils.copy(in, zos);
-                    }
-                    zos.closeEntry();
+            if (glob.isEmpty()) {
+                if (!root.supportQuickRecursiveListing()) {
+                    // avoid slow listing when the Glob can do a quicker job
+                    glob = "**";
                 }
             }
+            
+            if (glob.isEmpty()) {
+                Map<String, VirtualFile> nameToVirtualFiles = collectRecursivelyAllLegalChildren(dir);
+                sendZipUsingMap(zos, dir, nameToVirtualFiles);
+            } else {
+                Collection<String> listOfFile = dir.list(glob, null, /* TODO what is the user expectation? */true);
+                sendZipUsingListOfNames(zos, dir, listOfFile);
+            }
+        }
+    }
+
+    private static void sendZipUsingMap(ZipOutputStream zos, VirtualFile dir, Map<String, VirtualFile> nameToVirtualFiles) throws IOException {
+        for (Map.Entry<String, VirtualFile> entry : nameToVirtualFiles.entrySet()) {
+            String n = entry.getKey();
+
+            String relativePath;
+            // JENKINS-19947: traditional behavior is to prepend the directory name
+            relativePath = dir.getName() + '/' + n;
+
+            VirtualFile f = entry.getValue();
+            sendOneZipEntry(zos, f, relativePath);
+        }
+    }
+
+    private static void sendZipUsingListOfNames(ZipOutputStream zos, VirtualFile dir, Collection<String> listOfFileNames) throws IOException {
+        for (String relativePath : listOfFileNames) {
+            VirtualFile f = dir.child(relativePath);
+            sendOneZipEntry(zos, f, relativePath);
+        }
+    }
+
+    private static void sendOneZipEntry(ZipOutputStream zos, VirtualFile vf, String relativePath) throws IOException {
+        // In ZIP archives "All slashes MUST be forward slashes" (http://pkware.com/documents/casestudies/APPNOTE.TXT)
+        // TODO On Linux file names can contain backslashes which should not treated as file separators.
+        //      Unfortunately, only the file separator char of the master is known (File.separatorChar)
+        //      but not the file separator char of the (maybe remote) "dir".
+        ZipEntry e = new ZipEntry(relativePath.replace('\\', '/'));
+
+        e.setTime(vf.lastModified());
+        zos.putNextEntry(e);
+        try (InputStream in = vf.open()) {
+            IOUtils.copy(in, zos);
+        }
+        zos.closeEntry();
```

Comment:
add it to `finally`? Otherwise ZIP stream may be corrupted if `cf.open()` or copy fail

Reply:
Good improvement to the existing code